lib/periodicChecks.js

```diff
@@ -0,0 +1,183 @@
+ */
+
+const {
+  getAllOpenPullRequests,
+  hasPullRequestBeenApproved,
+  getChangelogLabelFromPullRequest,
+  getProjectOwnerFromLabel,
+  doesPullRequestHaveChangesRequested,
+  isUserAMemberOfTheOrganisation,
+} = require('./utils');
+const mergeConflictModule = require('./checkMergeConflicts');
+const pullRequestReviewModule = require('./checkPullRequestReview');
+const { teamLeads } = require('../userWhitelist.json');
+
+/**
+ *  This function checks if a pull request has merge conflicts.
+ *
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
+const hasMergeConflict = (pullRequest) => {
+  return pullRequest.mergeable === false;
+};
+
+/**
+ *  This function checks if a pull request has pending reviews.
+ *
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
+const hasPendingReviews = (pullRequest) => {
+  return pullRequest.requested_reviewers.length > 0;
+};
+
+/**
+ * This function pings and assigns all pending reviewers to a pull request.
+ * @param {import('probot').Context} context
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
+const pingAndAssignReviewers = async (context, pullRequest) => {
+  const reviewers = pullRequest.requested_reviewers.map(
+    (reviewer) => reviewer.login
+  );
+
+  context.github.issues.addAssignees(
+    context.repo({
+      issue_number: pullRequest.number,
+      assignees: reviewers,
+    })
+  );
+
+  context.github.issues.createComment(
+    context.repo({
+      issue_number: pullRequest.number,
+      body:
+        'Assigning @' +
+        reviewers.join(', @') +
+        ' for codeowner reviews. Thanks!',
+    })
+  );
+};
+
+/**
+ * This function assigns the PR author to a pull request.
+ * @param {import('probot').Context} context
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ * @param {string} comment
+ */
+const pingAndAssignPrAuthor = async (context, pullRequest, comment) => {
+  context.github.issues.addAssignees(
+    context.repo({
+      issue_number: pullRequest.number,
+      assignees: [pullRequest.user.login],
+    })
+  );
+  await context.github.issues.createComment(
+    context.repo({
+      issue_number: pullRequest.number,
+      body: comment,
+    })
+  );
+};
+
+/**
+ * This function ensures that a pull request is assigned.
+ * 1. Fetch all pull requests.
+ * 2. Filter out those without assignees.
+ * 3. Find out appropriate assignee.
+ *  - If pull request has a merge conflict    -- PR author.
+ *  - If pull request has pending reviews     -- Remaining reviewers.
+ *  - If pull request has changes requested   -- PR author.
+ *  - If pull request has been approved       -- Any member of the Organisation.
+ *  - If none of the above cases              -- Welfare team lead.
+ * 4. Assign appropriate assignee.
+ * 5. Ping appropriate assignee.
+ * @param {import('probot').Context} context
+ */
+const ensurePullRequestIsAssigned = async (context) => {
```

Comment:
So, what I would suggest is as follows:

Only have the `ensurePullRequestIsAssigned` function here and move all others to `pullRequestReview` lib or `mergeConflicts`. Just call the appropriate functions from here but let the other libs handle everything internally.

Basically the periodic check lib should be a blackbox which is just calling some checks periodically and they can be any checks but those checks should live in their respective modules.

Reply:
The `pingAndAssignReviewers` should still be in the periodic checks module, right?
Others have been moved.

Reply:
Yes, looks good now!